sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,641 @@
+
+            if (code != 0)
+            {
+                // Fail the entire build and push.  This includes the full output of the command so that at
+                // the end the user can review the full docker message about what the problem was.
+                //
+                // Note: a message about the command failing will have already been ephemerally reported to
+                // the status column.
+                throw new ResourceException($"{GetFailureMessage(cmd, args, reportFullCommandLine, code)}\n{stdout}", logResource);
+            }
+
+            return stdout;
+        }
+
+        /// <summary>
+        /// Runs a CLI command in a child process, returning a task for the process's exit. Both stdout
+        /// and stderr are redirected to process.stdout and process.stder by default.
+        /// </summary>
+        /// <param name="cmd">Command name to run.</param>
+        /// <param name="args">An array of command arguments.</param>
+        /// <param name="logResource">Used to specify the resource to associate command output with. Stderr messages
+        /// are always sent (since they may contain important information about something that's gone wrong).
+        /// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+        /// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+        /// content will be printed.</param>
+        /// <param name="reportFullCommandLine"></param>
+        /// <param name="reportErrorAsWarning"></param>
+        /// <param name="stdin">If defined, its contents are piped into stdin for the child process.</param>
+        /// <param name="env"></param>
+        /// <returns></returns>
+        private static async Task<(int, string)> RunCommandThatCanFail(
+            string cmd,
+            string[] args,
+            Resource logResource,
+            bool reportFullCommandLine,
+            bool reportErrorAsWarning,
+            string? stdin = null,
+            ImmutableDictionary<string, string>? env = null)
+        {
+            // Let the user ephemerally know the command we're going to execute.
+            Log.Info($"Executing {GetCommandLineMessage(cmd, args, reportFullCommandLine, env)}", logResource, ephemeral: true);
+
+            var process = new Process();
+            process.StartInfo.FileName = cmd;
+            process.StartInfo.Arguments = Utils.EscapeArguments(args);
+            process.StartInfo.RedirectStandardInput = stdin != null;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.RedirectStandardError = true;
+
+            try
+            {
+                process.Start();
+
+                if (stdin != null)
+                {
+                    process.StandardInput.Write(stdin);
+                    process.StandardInput.Close();
```

Comment:
interesting.  we need to close this?  why?

Reply:
otherwise the process might keep waiting for more input. I think there was a similar close in the node sdk